crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,343 @@ impl MeshletMesh {
-                    meshlet,
-                    mesh.attribute(Mesh::ATTRIBUTE_POSITION)
-                        .unwrap()
-                        .as_float3()
-                        .unwrap(),
-                )
-            })
-            .map(|bounds| MeshletBoundingSphere {
-                center: bounds.center.into(),
-                radius: bounds.radius,
+            .map(|meshlet| compute_meshlet_bounds(meshlet, &vertices))
+            .map(convert_meshlet_bounds)
+            .map(|bounding_sphere| MeshletBoundingSpheres {
+                self_culling: bounding_sphere,
+                self_lod: MeshletBoundingSphere {
+                    center: bounding_sphere.center,
+                    radius: 0.0,
+                },
+                parent_lod: MeshletBoundingSphere {
+                    center: bounding_sphere.center,
+                    radius: f32::MAX,
+                },
             })
-            .collect();
+            .collect::<Vec<_>>();
+        let worst_case_meshlet_triangles = meshlets
+            .meshlets
+            .iter()
+            .map(|m| m.triangle_count as u64)
+            .sum();
+
+        // Build further LODs
+        let mut simplification_queue = 0..meshlets.len();
+        let mut lod_level = 1;
+        while simplification_queue.len() > 1 {
+            // For each meshlet build a set of triangle edges
+            let triangle_edges_per_meshlet =
+                collect_triangle_edges_per_meshlet(simplification_queue.clone(), &meshlets);
+
+            // For each meshlet build a list of connected meshlets (meshlets that share a triangle edge)
+            let connected_meshlets_per_meshlet =
+                find_connected_meshlets(simplification_queue.clone(), &triangle_edges_per_meshlet);
+
+            // Group meshlets into roughly groups of 4, grouping meshlets with a high number of shared edges
+            // http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/manual.pdf
+            let groups = group_meshlets(
+                simplification_queue.clone(),
+                &connected_meshlets_per_meshlet,
+            );
+
+            let next_lod_start = meshlets.len();
+
+            for group_meshlets in groups.values().filter(|group| group.len() > 1) {
+                // Simplify the group to ~50% triangle count
+                let Some((simplified_group_indices, mut group_error)) =
+                    simplify_meshlet_groups(group_meshlets, &meshlets, &vertices, lod_level)
+                else {
+                    continue;
+                };
+
+                // Add the maximum child error to the parent error to make parent error cumulative from LOD 0
+                // (we're currently building the parent from its children)
+                group_error += group_meshlets.iter().fold(group_error, |acc, meshlet_id| {
+                    acc.max(bounding_spheres[*meshlet_id].self_lod.radius)
+                });
```

Comment:
At some point we might want to consider importing `itertools`, as using `max` from that crate would be easier to read.

Reply:
I already use itertools elsewhere in this file, I'll use it.

Reply:
Won't work on f32.